### PR TITLE
Release restriction in `OneHotToNumeric` that the categoricals are the trailing dimensions

### PR DESCRIPTION
**Pull request by jduerholt**
## Motivation

This PR refactors the `OneHotToNumeric` `InputTransform` in a way that it is not needed anymore that the categoricals are the trailing dimensions. Thus, making the input transform more flexible.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Unit tests.

### CODE REVIEW

**Review by saitcakmak — APPROVED** [review]
Thanks for supporting more general use cases!

**Comment by jduerholt** [comment]
Hi @saitcakmak,

thanks for reviewing. It looks to me that the PR causes some problem with Meta's internal tests, which I cannot see. Can you give me a hint on what is causing the problem(s) so that I can proceed with fixing?

Best,

Johannes

**Comment by Balandat** [comment]
@jduerholt thanks for the concern here, but looks like there isn't actually anything wrong. These seem to be some warnings about some tests not running or timing out (which is almost unavoidable given that there are tens of thousands of internal tests running), but somehow those get flagged as "unsuccessful" here on GitHub. 

Thanks for your contribution, btw :)

**Comment by saitcakmak** [comment]
For more context: After pulling in the changes from GitHub, we need approval from a second reviewer internally. I'm waiting on that internal review before I can merge this in. Nothing needed from your end!

File: test/models/transforms/test_input.py
Comment (esantorella) [inline test/models/transforms/test_input.py:988]: Thanks for specifying the message!

File: test/models/transforms/test_input.py
Comment (jduerholt) [reply test/models/transforms/test_input.py:986]: I am not sure about this `subTest` thingy, never used it as I am more familiar with `pytest` than`unittest`, so you would split the whole test method into subtests or only the first checks that the exceptions are raised?

For me, the lenght of the test is fine, but if you want I can change it ;)

**Pull request closed without merging**
